<launch>
    <arg name="robot_name"             default="argos"/> <!-- Change this for namespacing. -->
    <arg name="rviz"                   default="false"/> <!-- Set to true to run rviz in parallel. -->
    <arg name="lite"                   default="false" />  <!-- Set to true if you're using CHAMP lite version. Only useful for microcontrollers. -->
    <arg name="ros_control_file"       default="$(find argos_config)/config/ros_control/ros_control.yaml" /> <!-- Path to ROS Control configurations. Do not touch. -->
    <arg name="gazebo_world"           default="$(find argos_config)/worlds/arena_world.world" /> <!-- Path to Gazebo world you want to load. -->
    <arg name="gui"                    default="true"/>
    <arg name="paused"                 default="false"/>
    <arg name="world_init_x"           default="0.0" /> <!-- X Initial position of the robot in Gazebo World -->
    <arg name="world_init_y"           default="0.0" /> <!-- Y Initial position of the robot in Gazebo World -->
    <arg name="world_init_z"           default="1.08" />
    <arg name="world_init_heading"     default="0.0" /> <!-- Initial heading of the robot in Gazebo World -->
    <arg name="publish_joint_control"  default="true"/>
    <arg name="world_frame"            default="argos/world"/>
    <arg name="namespace"              default="argos"/>
    <arg name="child_frame_id"         default="argos/base_link"/>
    <arg name="imu_topic"              default="argos/imu/data"/>
    <arg name="ground_truth"           default="false"/>
    <arg name="ptl"                    default="false"/>  <!-- pointcloud to laserscan-->
    <arg name="dtl"                    default="true"/>   <!-- disparity to laserscan-->
    
    
    

    <param name="use_sim_time" value="true" />

    <include file="$(find argos_config)/launch/bringup.launch">
        <arg name="robot_name"             value="$(arg robot_name)"/>
        <arg name="gazebo"                 value="true"/>
        <arg name="lite"                   value="$(arg lite)"/>
        <arg name="rviz"                   value="$(arg rviz)"/>
        <arg name="joint_controller_topic" value="joint_group_position_controller/command"/>
        <arg name="hardware_connected"     value="false"/>
        <arg name="publish_foot_contacts"  value="true"/>
        <arg name="publish_joint_control"  value="$(arg publish_joint_control)"/>
        <arg name="close_loop_odom"        value="true"/>
        <arg name="ground_truth"           value="$(arg ground_truth)"/>
    </include>

    <include file="$(find champ_gazebo)/launch/gazebo.launch">
        <arg name="robot_name"         value="$(arg robot_name)"/>
        <arg name="lite"               value="$(arg lite)"/>
        <arg name="ros_control_file"   value="$(arg ros_control_file)"/>
        <arg name="gazebo_world"       value="$(arg gazebo_world)"/>
        <arg name="world_init_x"       value="$(arg world_init_x)" />
        <arg name="world_init_y"       value="$(arg world_init_y)" />
        <arg name="world_init_z"       value="$(arg world_init_z)" />
        <arg name="world_init_heading" value="$(arg world_init_heading)" />
        <arg name="gui"                value="$(arg gui)" />
        <arg name="paused"             value="$(arg paused)"/>
    </include>    


    <node ns="argos/camera" pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc">
        <param name="prefilter_size"            value="5"/>
        <param name="prefilter_cap"             value="25"/>
        <param name="correlation_window_size"   value="21"/>
        <param name="min_disparity"             value="-59"/>
        <param name="disparity_range"           value="224"/>
        <param name="uniqueness_ratio"          value="69.0"/>
        <param name="texture_threshold"         value="10"/>
        <param name="speckle_size"              value="0"/>
        <param name="speckle_range"             value="2"/>
    </node>
    <group if="$(arg ground_truth)">
        <node name="odom_to_tf" pkg="message_to_tf" type="message_to_tf" args="argos/odom/ground_truth">
            <param name="~frame_id"                 value="$(arg world_frame)" />
            <param name="~stabilized_frame_id"      value="/$(arg namespace)/base_stabilized" />
            <param name="~footprint_frame_id"       value="/$(arg namespace)/base_footprint" />
            <param name="~child_frame_id"           value="$(arg child_frame_id)" />
            <param name="~imu_topic"                value="$(arg imu_topic)"/>
        </node>
        
        <node pkg="tf" type="static_transform_publisher" name ="map_to_world" args="0 0 0 0 0 0 argos/map argos/world 100" />
    </group>

    <group if="$(arg dtl)">
        <group unless="$(arg ptl)">
            <include file="$(find argos_config)/launch/disparity_to_depth.launch">
                <arg name="camera_name"                value="/argos/camera"/>
            </include>

            <include file="$(find argos_config)/launch/depth_to_laser.launch">
                <arg name="robot_namespace"            value="argos"/>
                <arg name="camera_name"                value="argos/camera"/>
                <arg name="camera_link"                value="argos/camera_center"/>
            </include>
        </group>
    </group>

    <group if="$(arg ptl)">
        <group ns="argos">
            <include file="$(find argos_config)/launch/pointcloud_to_laserscan/point_to_laser_node.launch" >
                <arg name="camera_name"             value="camera"/>
                <arg name="target_frame"            value="argos/camera_center"/>
                <arg name="min_height"              value="-1.0"/>
            </include>
        </group>
    </group>


</launch>
